### ISSUE TITLE
Repeatedly issues with signing

### ISSUE THREAD

**Issue by dnperfors** [jabbera/vsts-authenticode#20]
Hi,

Since a couple of weeks we are using this plugin in our builds in order to be able to use the Secure file storage of TFS.
Most of the time this works like a charm, but sometimes we notice that we get the following message:

> 2019-03-18T07:48:27.7364419Z [command]C:\BuildAgent\_work\_tasks\authenticode-sign_752fe535-ed47-4c2c-afcf-0778adb0bb12\3.0.0\x64\signtool.exe sign /tr http://timestamp.digicert.com /td sha256 /f C:\BuildAgent\_work\_temp\CodeSignCertificate-2022-02-20.pfx /p *** /fd sha256 <...several assembly files...>
> 2019-03-18T07:48:27.9965373Z SignTool Error: An error occurred while attempting to load the signing
> 2019-03-18T07:48:27.9966085Z 
> 2019-03-18T07:48:27.9966955Z 	certificate from: C:\BuildAgent\_work\_temp\CodeSignCertificate-2022-02-20.pfx
> 2019-03-18T07:48:27.9967702Z 
> 2019-03-18T07:48:28.0134499Z Error attempting to sign. Attempt number: 0. Exception text: Error: C:\BuildAgent\_work\_tasks\authenticode-sign_752fe535-ed47-4c2c-afcf-0778adb0bb12\3.0.0\x64\signtool.exe failed with return code: 1
> 2019-03-18T07:48:28.0140299Z Sleeping for 15 second(s)

How could we debug these kind of issue and get the root cause of it?

Regards,
David

**Comment by jabbera** [comment]
Hi,

Can you make sure system.debug is set to true and post the log of the task please. 

Do all the retries fail with the same pattern? There is a loop of retries meant to deal with timestamp server issues, and I’m surprised any transient error isn’t being handled by this logic.

**Comment by dnperfors** [comment]
I enable debug logging for the build, but it could take some time before I have the information. It doesn't run consistently into this issue.

it does fail for 5 retries, because the build breaks, I also tried other timestamp servers, but looking at the error message I don't really think it is caused by that, but rather some weird issue with getting the certificate itself?

**Comment by dnperfors** [comment]
I tried to get more information by passing `/debug` as additional argument, but it doesn't give any additional output.

The strange thing is that we have a separate build definition for another branch which is constantly passing, while this build definition on our master branch is constantly failing... They use exactly the same certificate, agents, etc.

**Comment by pswetz** [comment]
We also experience this issue quite a bit. Sometimes everything is fine, other times we end up with this error in multiple pipelines.

##[debug]Evaluating condition for step: 'Authenticode Sign engine'
##[debug]Evaluating: SucceededNode()
##[debug]Evaluating SucceededNode:
##[debug]=> True
##[debug]Result: True
##[section]Starting: Authenticode Sign engine
==============================================================================
Task         : Authenticode Sign
Description  : Authenticode sign executable code.
Version      : 3.0.0
Author       : Michael Barry
Help         : 
==============================================================================
##[debug]agent.TempDirectory=D:\a\_temp
##[debug]loading inputs and endpoints
##[debug]loading ENDPOINT_AUTH_PARAMETER_SYSTEMVSSCONNECTION_ACCESSTOKEN
##[debug]loading ENDPOINT_AUTH_SCHEME_SYSTEMVSSCONNECTION
##[debug]loading ENDPOINT_AUTH_SYSTEMVSSCONNECTION
##[debug]loading INPUT_CERTIFICATELOCATION
##[debug]loading INPUT_CERTIFICATESELECTIONMETHOD
##[debug]loading INPUT_FILEALGO
##[debug]loading INPUT_FILEPATH
##[debug]loading INPUT_PFXFILE
##[debug]loading INPUT_PFXPASSWORD
##[debug]loading INPUT_PFXSECUREFILE
##[debug]loading INPUT_RETRYCOUNT
##[debug]loading INPUT_SIGNROOTPATH
##[debug]loading INPUT_TIMESTAMPALGO
##[debug]loading INPUT_TIMESTAMPSERVER
##[debug]loading INPUT_TIMESTAMPSERVERDELAY
##[debug]loading SECRET_PROD-CERT-PFX-KEY-PWD
##[debug]loading SECRET_SYSTEM_ACCESSTOKEN
##[debug]loading SECRET_TEST-CERT-PFX-KEY-PWD
##[debug]loading SECUREFILE_TICKET_0d25dafe-1442-4810-8be3-ce6625abc914
##[debug]loaded 19
##[debug]Agent.ProxyUrl=undefined
##[debug]Agent.CAInfo=undefined
##[debug]Agent.ClientCert=undefined
##[debug]signToolLocation=null
##[debug]Using tool location: D:\a\_tasks\authenticode-sign_xxxxxxxxx\3.0.0\x64\signtool.exe
##[debug]check path : D:\a\_tasks\authenticode-sign_xxxxxxxx\3.0.0\x64\signtool.exe
##[debug]retryCount=5
##[debug]timestampServerDelay=15
##[debug]which 'D:\a\_tasks\authenticode-sign_xxxxxx\3.0.0\x64\signtool.exe'
##[debug]found: 'D:\a\_tasks\authenticode-sign_xxxxxx\3.0.0\x64\signtool.exe'
##[debug]which 'D:\a\_tasks\authenticode-sign_xxxxxx\3.0.0\x64\signtool.exe'
##[debug]found: 'D:\a\_tasks\authenticode-sign_xxxxxxx\3.0.0\x64\signtool.exe'
##[debug]timestampServer=http://timestamp.digicert.com
##[debug]timestampAlgo=sha256
##[debug]certificateLocation=secureFile
##[debug]pfxSecureFile=xxxxxxxxxxxx
##[debug]System.TeamFoundationCollectionUri=https://xxxxxxxx.visualstudio.com/
##[debug]SYSTEMVSSCONNECTION auth param ACCESSTOKEN = ***
##[debug]secure file name for id xxxx = acmetestcert.pfx
##[debug]Agent.TempDirectory=D:\a\_temp
##[debug]Absolute path for pathSegments: D:\a\_temp,acmetestcert.pfx = D:\a\_temp\acmetestcert.pfx
##[debug]Downloading secure file contents to: D:\a\_temp\acmetestcert.pfx
##[debug]SYSTEM.TEAMPROJECT=Pathfinder
##[debug]secure file ticket for id xxxxxxx = ***
##[debug]Downloaded secure file contents to: D:\a\_temp\acmetestcert.pfx
##[debug]check path : D:\a\_temp\acmetestcert.pfx
##[debug]pfxPassword=***
##[debug]additionalArguments=null
##[debug]fileAlgo=sha256
##[debug]filePath=engine.exe
##[debug]signRootPath=D:\a\1\s\source\xxxxxx\bin\Release-Dev
##[debug]defaultRoot: 'D:\a\1\s\source\xxxxx\bin\Release-Dev'
##[debug]findOptions.followSpecifiedSymbolicLink: 'true'
##[debug]findOptions.followSymbolicLinks: 'true'
##[debug]matchOptions.debug: 'false'
##[debug]matchOptions.nobrace: 'true'
##[debug]matchOptions.noglobstar: 'undefined'
##[debug]matchOptions.dot: 'true'
##[debug]matchOptions.noext: 'undefined'
##[debug]matchOptions.nocase: 'true'
##[debug]matchOptions.nonull: 'undefined'
##[debug]matchOptions.matchBase: 'undefined'
##[debug]matchOptions.nocomment: 'undefined'
##[debug]matchOptions.nonegate: 'undefined'
##[debug]matchOptions.flipNegate: 'undefined'
##[debug]pattern: 'engine.exe'
##[debug]findPath: 'D:\a\1\s\source\PosApiBootstrapper\bin\Release-Dev\engine.exe'
##[debug]statOnly: 'true'
##[debug]found 1 paths
##[debug]applying include pattern
##[debug]adjustedPattern: 'D:\a\1\s\source\PosApiBootstrapper\bin\Release-Dev\engine.exe'
##[debug]1 matches
##[debug]1 final results
[command]D:\a\_tasks\authenticode-sign_xxxxxxxx\3.0.0\x64\signtool.exe sign /tr http://timestamp.digicert.com /td sha256 /f D:\a\_temp\acmetestcert.pfx /p *** /fd sha256 D:\a\1\s\source\xxxxxxx\bin\Release-Dev\engine.exe
SignTool Error: An error occurred while attempting to load the signing

	certificate from: D:\a\_temp\acmetestcert.pfx

Error attempting to sign. Attempt number: 0. Exception text: Error: D:\a\_tasks\authenticode-sign_xxxxxx\3.0.0\x64\signtool.exe failed with return code: 1
Sleeping for 15 second(s)
[command]D:\a\_tasks\authenticode-sign_xxxxxxx\3.0.0\x64\signtool.exe sign /tr http://timestamp.digicert.com /td sha256 /f D:\a\_temp\acmetestcert.pfx /p *** /fd sha256 D:\a\1\s\source\xxxxxxx\bin\Release-Dev\engine.exe
SignTool Error: An error occurred while attempting to load the signing

	certificate from: D:\a\_temp\acmetestcert.pfx

Error attempting to sign. Attempt number: 1. Exception text: Error: D:\a\_tasks\authenticode-sign_xxxxxxxx\3.0.0\x64\signtool.exe failed with return code: 1
Sleeping for 15 second(s)
[command]D:\a\_tasks\authenticode-sign_xxxxxxx\3.0.0\x64\signtool.exe sign /tr http://timestamp.digicert.com /td sha256 /f D:\a\_temp\acmetestcert.pfx /p *** /fd sha256 D:\a\1\s\source\xxxxxxxxx\bin\Release-Dev\engine.exe
SignTool Error: An error occurred while attempting to load the signing

	certificate from: D:\a\_temp\acmetestcert.pfx

Error attempting to sign. Attempt number: 2. Exception text: Error: D:\a\_tasks\authenticode-sign_752fe535-ed47-4c2c-afcf-0778adb0bb12\3.0.0\x64\signtool.exe failed with return code: 1
Sleeping for 15 second(s)
[command]D:\a\_tasks\authenticode-sign_752fe535-ed47-4c2c-afcf-0778adb0bb12\3.0.0\x64\signtool.exe sign /tr http://timestamp.digicert.com /td sha256 /f D:\a\_temp\acmetestcert.pfx /p *** /fd sha256 D:\a\1\s\source\PosApiBootstrapper\bin\Release-Dev\engine.exe
SignTool Error: An error occurred while attempting to load the signing

	certificate from: D:\a\_temp\acmetestcert.pfx

Error attempting to sign. Attempt number: 3. Exception text: Error: D:\a\_tasks\authenticode-sign_xxxxxxx\3.0.0\x64\signtool.exe failed with return code: 1
Sleeping for 15 second(s)
[command]D:\a\_tasks\authenticode-sign_xxxxxxx\3.0.0\x64\signtool.exe sign /tr http://timestamp.digicert.com /td sha256 /f D:\a\_temp\acmetestcert.pfx /p *** /fd sha256 D:\a\1\s\source\xxxxxxx\bin\Release-Dev\engine.exe
SignTool Error: An error occurred while attempting to load the signing

	certificate from: D:\a\_temp\acmetestcert.pfx

Error attempting to sign. Attempt number: 4. Exception text: Error: D:\a\_tasks\authenticode-sign_752fe535-ed47-4c2c-afcf-0778adb0bb12\3.0.0\x64\signtool.exe failed with return code: 1
Sleeping for 15 second(s)
##[debug]task result: Failed
##[error]Unable to sign Error: D:\a\_tasks\authenticode-sign_xxxxxxx\3.0.0\x64\signtool.exe failed with return code: 1
##[debug]Processed: ##vso[task.issue type=error;]Unable to sign Error: D:\a\_tasks\authenticode-sign_xxxxxxx\3.0.0\x64\signtool.exe failed with return code: 1
##[debug]Processed: ##vso[task.complete result=Failed;]Unable to sign Error: D:\a\_tasks\authenticode-sign_xxxxxxxx\3.0.0\x64\signtool.exe failed with return code: 1
##[debug]task result: Failed
##[error]Unable to sign Error: D:\a\_tasks\authenticode-sign_752fe535-ed47-4c2c-afcf-0778adb0bb12\3.0.0\x64\signtool.exe failed with return code: 1
##[debug]Processed: ##vso[task.issue type=error;]Unable to sign Error: D:\a\_tasks\authenticode-sign_752fe535-ed47-4c2c-afcf-0778adb0bb12\3.0.0\x64\signtool.exe failed with return code: 1
##[debug]Processed: ##vso[task.complete result=Failed;]Unable to sign Error: D:\a\_tasks\authenticode-sign_752fe535-ed47-4c2c-afcf-0778adb0bb12\3.0.0\x64\signtool.exe failed with return code: 1
##[debug]secure file name for id 0d25dafe-1442-4810-8be3-ce6625abc914 = acmetestcert.pfx
##[debug]Agent.TempDirectory=D:\a\_temp
##[debug]Absolute path for pathSegments: D:\a\_temp,acmetestcert.pfx = D:\a\_temp\acmetestcert.pfx
##[debug]Deleting secure file at: D:\a\_temp\acmetestcert.pfx
##[debug]rm -rf D:\a\_temp\acmetestcert.pfx
##[debug]removing file
(node:2532) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: D:\a\_tasks\authenticode-sign_xxxxxxxxxx\3.0.0\x64\signtool.exe failed with return code: 1
##[section]Finishing: Authenticode Sign engine

**Comment by jabbera** [comment]
Based on this new log I wonder if the pfx isn't being downloaded properly. Can you try subbing out my home grown secure file feature with the official one?

https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/utility/download-secure-file?view=azure-devops

**Comment by dnperfors** [comment]
I tried to use the `download-secure-file` tasks and the first try did work.. I will change some other builds, and make sure it consistently works.

Funny enough, the download-secure-file doesn't show up in the the log at all...

**Comment by zauggm** [comment]
I have also been experiencing this issue and can reproduce it quite easily.  We have some builds that take quite a bit of time to complete and that seems to be what causes the issue. I've found that any signing step that tries to execute more than 20 minutes after the start of the build will fail with the error that other posters provided above. I have a build that floats around that 20 minute time range. If the signing step occurs before 20 minutes it succeeds. If the signing step occurs after 20 minutes it fails.

I did try using official download secure file task in my pipeline but it did not solve the issue.  

Any thoughts on if this 20 minute timeout could be within the authenticode task?
